src/effects/native/echoeffect.cpp

```diff
@@ -143,90 +144,98 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
-                SampleUtil::clampSample(gs.delay_buf[gs.write_position + 1]);
-        INCREMENT_RING(gs.write_position, EchoGroupState::kChannelCount, delay_samples);
-    }
+        INCREMENT_RING(read_position, EchoGroupState::kChannelCount,
+                gs.delay_buf.size());
 
-    // Pingpong the output.  If the pingpong value is zero, all of the
-    // math below should result in a simple copy of delay buf to pOutput.
-    for (unsigned int i = 0; i < numSamples; i += EchoGroupState::kChannelCount) {
-        if (gs.ping_pong_left) {
+        // Actual delays distort and saturate, so clamp the buffer here.
+        gs.delay_buf[gs.write_position] = SampleUtil::clampSample(
+                pInput[i] * send_ramped +
+                bufferedSample0 * feedback_ramped);
+        gs.delay_buf[gs.write_position + 1] = SampleUtil::clampSample(
+                pInput[i + 1] * send_ramped +
+                bufferedSample0 * feedback_ramped);
+
+        // Pingpong the output.  If the pingpong value is zero, all of the
+        // math below should result in a simple copy of delay buf to pOutput.
+        if (gs.ping_pong < delay_samples / 2) {
             // Left sample plus a fraction of the right sample, normalized
             // by 1 + fraction.
             pOutput[i] = pInput[i] +
-                    ((gs.delay_buf[read_position] +
-                            gs.delay_buf[read_position + 1] * pingpong_frac) /
-                    (1 + pingpong_frac)) / 2.0;
+                    ((bufferedSample0 + bufferedSample1 * pingpong_frac) /
+                    (1 + pingpong_frac));
             // Right sample reduced by (1 - fraction)
             pOutput[i + 1] = pInput[i + 1] +
-                    (gs.delay_buf[read_position + 1] * (1 - pingpong_frac)) / 2.0;
+                    (bufferedSample1 * (1 - pingpong_frac));
         } else {
             // Left sample reduced by (1 - fraction)
             pOutput[i] = pInput[i] +
-                    (gs.delay_buf[read_position] * (1 - pingpong_frac)) / 2.0;
+                    (bufferedSample0 * (1 - pingpong_frac));
             // Right sample plus fraction of left sample, normalized by
             // 1 + fraction
             pOutput[i + 1] = pInput[i + 1] +
-                    ((gs.delay_buf[read_position + 1] +
-                            gs.delay_buf[read_position] * pingpong_frac) /
-                    (1 + pingpong_frac)) / 2.0;
+                    ((bufferedSample1 + bufferedSample0 * pingpong_frac) /
+                    (1 + pingpong_frac));
         }
 
-        INCREMENT_RING(read_position, EchoGroupState::kChannelCount, delay_samples);
-        // If the buffer has looped around, flip-flop the ping-pong.
-        if (read_position == 0) {
-            gs.ping_pong_left = !gs.ping_pong_left;
+        INCREMENT_RING(gs.write_position, EchoGroupState::kChannelCount,
+                gs.delay_buf.size());
+
+        if (++gs.ping_pong >= delay_samples) {
```

Comment:
I think combining variable assignment and condition testing on the same line is unnecessarily hard to read. Please split into two lines like:
```
++gs.ping_pong;
if (gs.ping_pong >= delay_samples) {
```